### PR TITLE
(Discord RPC) Migrate to discord-rpc-csharp

### PR DESCRIPTION
**Pull request by shatyuka**
Major changes:
- Migrate from official Discord Game SDK to [discord-rpc-csharp](https://github.com/CollapseLauncher/Hi3Helper.SharpDiscordRPC)
- Fix hide idle status
- Auto reconnect and send presence to Discord

### CODE REVIEW

File: CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPage.xaml.cs [hunk 1]
```diff
@@ -154,12 +153,13 @@ private void InitializeSettings(object sender, RoutedEventArgs e)
                     PageContent.Visibility = Visibility.Collapsed;
                     OverlayTitle.Text = Lang._GameSettingsPage.OverlayNotInstalledTitle;
                     OverlaySubtitle.Text = Lang._GameSettingsPage.OverlayNotInstalledSubtitle;
-
-                    return;
                 }
+                else
```
Comment (bagusnl) [inline CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPage.xaml.cs:157]: I dont think its necessary to make a new else block just for this and we can just use the else if block above it

File: CollapseLauncher/XAMLs/MainApp/Pages/GenshinGameSettingsPage.xaml.cs [hunk 2]
```diff
@@ -189,12 +188,13 @@ private void InitializeSettings(object sender, RoutedEventArgs e)
                     PageContent.Visibility = Visibility.Collapsed;
                     OverlayTitle.Text = Lang._StarRailGameSettingsPage.OverlayNotInstalledTitle;
                     OverlaySubtitle.Text = Lang._StarRailGameSettingsPage.OverlayNotInstalledSubtitle;
-
-                    return;
                 }
+                else
```
Comment (bagusnl) [inline CollapseLauncher/XAMLs/MainApp/Pages/GenshinGameSettingsPage.xaml.cs:192]: same here

File: CollapseLauncher/XAMLs/MainApp/Pages/RepairPage.xaml.cs [hunk 3]
```diff
@@ -216,6 +213,12 @@ private void InitializeLoaded(object sender, RoutedEventArgs e)
                 OverlayTitle.Text = Lang._GameRepairPage.OverlayGameRunningTitle;
                 OverlaySubtitle.Text = Lang._GameRepairPage.OverlayGameRunningSubtitle;
             }
+            else
```
Comment (bagusnl) [inline CollapseLauncher/XAMLs/MainApp/Pages/RepairPage.xaml.cs:216]: same here

File: CollapseLauncher/XAMLs/MainApp/Pages/StarRailGameSettingsPage.xaml.cs [hunk 4]
```diff
@@ -167,12 +166,13 @@ private void InitializeSettings(object sender, RoutedEventArgs e)
                     PageContent.Visibility = Visibility.Collapsed;
                     OverlayTitle.Text = Lang._StarRailGameSettingsPage.OverlayNotInstalledTitle;
                     OverlaySubtitle.Text = Lang._StarRailGameSettingsPage.OverlayNotInstalledSubtitle;
-
-                    return;
                 }
```
Comment (bagusnl) [inline CollapseLauncher/XAMLs/MainApp/Pages/StarRailGameSettingsPage.xaml.cs:169]: same here

**Review by bagusnl — COMMENTED** [review]
General early code feedback <3. Also as the #398 is merged, all GSP modules have been moved, please adjust to the change

File: CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPage.xaml.cs (hunk 1 above)
Comment (shatyuka) [reply CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPage.xaml.cs:157]: Set the presence only if the game is not running and not broken.

**Comment by bagusnl** [comment]
Found a bug where when idle status is disabled, changing RPC activity to other than idle not bringing up the RPC up

**Comment by shatyuka** [comment]
> Found a bug where when idle status is disabled, changing RPC activity to other than idle not bringing up the RPC up

Seems like a multi-thread issue. Delaying for a while is not reliable.

**Comment by bagusnl** [comment]
While the initial issue is fixed, it seems that now when "Show RPC when idling" is disabled, the RPC is not properly disposed and stays active. This can be fixed by adding check if `_presence` is null on `UpdateActivity()` then do a `Dispose()`.

Doing this will require you to also do a check if the `_client == null` also in `UpdateActivity()` to properly brought up the RPC back up again and avoid crash while also doing check in the `EnablePresence()` to avoid double RPC issue

While that is pretty straightforward, the one issue that I'm still wondering since yesterday is "Activity cleared" log spam that happened from this line https://github.com/CollapseLauncher/Collapse/blob/79ae950267ae386678175843754a239153e65f4e/Hi3Helper.Core/Classes/DiscordPresence/DiscordPresenceManager.cs#L110-#113
image

This happens when "Show RPC while idling" is enable and you are on homepage. I do wonder if its just my use of [arRPC](https://github.com/OpenAsar/arRPC) instead of normal client that gives the issue or what i dont fully yet to understand. 

Though other than those 2 points that I brought up, it works all great on the migration, good job ! <3

**Comment by shatyuka** [comment]
> While the initial issue is fixed, it seems that now when "Show RPC when idling" is disabled, the RPC is not properly disposed and stays active. This can be fixed by adding check if `_presence` is null on `UpdateActivity()` then do a `Dispose()`.

I did this on purpose. `UpdateActivity` with a null presence actually clears current rich presence. See [ClearPresence](https://github.com/CollapseLauncher/Hi3Helper.SharpDiscordRPC/blob/a6225eab1e94c79f861e248610cd98dbd82d6123/DiscordRPC/DiscordRpcClient.cs#L842-L858).
A long connection will be faster than on-demand connection as I think. This behaves like hiding the presence temporarily, rather than disabling it.

**Comment by bagusnl** [comment]
> I did this on purpose. `UpdateActivity` with a null presence actually clears current rich presence. See [ClearPresence](https://github.com/CollapseLauncher/Hi3Helper.SharpDiscordRPC/blob/a6225eab1e94c79f861e248610cd98dbd82d6123/DiscordRPC/DiscordRpcClient.cs#L842-L858).

I see, though under my system it does not clear the RPC status and kept the last one instead

https://github.com/CollapseLauncher/Collapse/assets/28079733/17f8d367-a7e7-4dae-ac7c-848f383869c1

**Comment by shatyuka** [comment]
There shouldn't be so much "Activity cleared!". Seems that OnPresenceUpdate event message is null? Try restarting Discord maybe.

**Comment by bagusnl** [comment]
I did already that and the issue persist. But it could be arRPC issue, if you see no issues with both "Activity cleared!" log spam and the idle presence disablement then we can merge this for now (asking others to test it) and I'll send issue ticket with arRPC devs about this

Also I see a lot of Trim Warnings, note for self to either fix this or disable trim to the assembly until it gets fixed
<details>
  <summary>Trim warnings</summary>

  ```csharp
D:\a\Collapse\Collapse\Hi3Helper.SharpDiscordRPC\DiscordRPC\RPC\Payload\PayloadArgument.cs(44,13): Trim analysis warning IL2026: DiscordRPC.RPC.Payload.ArgumentPayload.GetObject<T>(): Using member 'System.Text.Json.JsonSerializer.Deserialize<TValue>(JsonDocument, JsonSerializerOptions)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonTypeInfo or JsonSerializerContext, or make sure all of the required types are preserved. [D:\a\Collapse\Collapse\CollapseLauncher\CollapseLauncher.csproj]
D:\a\Collapse\Collapse\Hi3Helper.SharpDiscordRPC\DiscordRPC\RPC\Payload\PayloadArgument.cs(34,13): Trim analysis warning IL2026: DiscordRPC.RPC.Payload.ArgumentPayload.SetObject(Object): Using member 'System.Text.Json.JsonSerializer.SerializeToDocument<TValue>(TValue, JsonSerializerOptions)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonTypeInfo or JsonSerializerContext, or make sure all of the required types are preserved. [D:\a\Collapse\Collapse\CollapseLauncher\CollapseLauncher.csproj]
D:\a\Collapse\Collapse\Hi3Helper.SharpDiscordRPC\DiscordRPC\Converters\EnumSnakeCaseConverter.cs(72,13): Trim analysis warning IL2070: DiscordRPC.Converters.EnumSnakeCaseConverter.TryParseEnum(Type, String, Object&): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors', 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.PublicFields', 'DynamicallyAccessedMemberTypes.PublicNestedTypes', 'DynamicallyAccessedMemberTypes.PublicProperties', 'DynamicallyAccessedMemberTypes.PublicEvents' in call to 'System.Type.GetMembers(BindingFlags)'. The parameter 'enumType' of method 'DiscordRPC.Converters.EnumSnakeCaseConverter.TryParseEnum(Type, String, Object&)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [D:\a\Collapse\Collapse\CollapseLauncher\CollapseLauncher.csproj]
D:\a\Collapse\Collapse\Hi3Helper.SharpDiscordRPC\DiscordRPC\Converters\EnumSnakeCaseConverter.cs(72,13): Trim analysis warning IL2075: DiscordRPC.Converters.EnumSnakeCaseConverter.TryParseEnum(Type, String, Object&): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors', 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.PublicFields', 'DynamicallyAccessedMemberTypes.PublicNestedTypes', 'DynamicallyAccessedMemberTypes.PublicProperties', 'DynamicallyAccessedMemberTypes.PublicEvents' in call to 'System.Type.GetMembers(BindingFlags)'. The return value of method 'System.Linq.Enumerable.First<TSource>(IEnumerable<TSource>)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [D:\a\Collapse\Collapse\CollapseLauncher\CollapseLauncher.csproj]
D:\a\Collapse\Collapse\Hi3Helper.SharpDiscordRPC\DiscordRPC\Converters\EnumSnakeCaseConverter.cs(34,13): Trim analysis warning IL2075: DiscordRPC.Converters.EnumSnakeCaseConverter.Write(Utf8JsonWriter, Object, JsonSerializerOptions): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors', 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.PublicFields', 'DynamicallyAccessedMemberTypes.PublicNestedTypes', 'DynamicallyAccessedMemberTypes.PublicProperties', 'DynamicallyAccessedMemberTypes.PublicEvents' in call to 'System.Type.GetMembers(BindingFlags)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [D:\a\Collapse\Collapse\CollapseLauncher\CollapseLauncher.csproj]
D:\a\Collapse\Collapse\Hi3Helper.SharpDiscordRPC\DiscordRPC\RPC\Payload\PayloadEvent.cs(44,13): Trim analysis warning IL2026: DiscordRPC.RPC.Payload.EventPayload.GetObject<T>(): Using member 'System.Text.Json.JsonSerializer.Deserialize<TValue>(JsonDocument, JsonSerializerOptions)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonTypeInfo or JsonSerializerContext, or make sure all of the required types are preserved. [D:\a\Collapse\Collapse\CollapseLauncher\CollapseLauncher.csproj]
D:\a\Collapse\Collapse\Hi3Helper.SharpDiscordRPC\DiscordRPC\IO\PipeFrame.cs(103,13): Trim analysis warning IL2026: DiscordRPC.IO.PipeFrame.GetObject<T>(): Using member 'System.Text.Json.JsonSerializer.Deserialize<TValue>(String, JsonSerializerOptions)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonTypeInfo or JsonSerializerContext, or make sure all of the required types are preserved. [D:\a\Collapse\Collapse\CollapseLauncher\CollapseLauncher.csproj]
D:\a\Collapse\Collapse\Hi3Helper.SharpDiscordRPC\DiscordRPC\IO\PipeFrame.cs(80,13): Trim analysis warning IL2026: DiscordRPC.IO.PipeFrame.SetObject(Object): Using member 'System.Text.Json.JsonSerializer.Serialize<TValue>(TValue, JsonSerializerOptions)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonTypeInfo or JsonSerializerContext, or make sure all of the required types are preserved. [D:\a\Collapse\Collapse\CollapseLauncher\CollapseLauncher.csproj]
  ```
</details>

**Review by Cryotechnic — APPROVED** [review]
Can confirm that there are no issues with activity clearing when idling when using Discord Stable. I think we can resolve merge conflict and merge.

**Pull request merged**